lmdeploy/pytorch/engine/engine.py

```diff
@@ -0,0 +1,866 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import time
+from dataclasses import dataclass
+from queue import Queue
+from threading import Thread
+from typing import Any, Dict, List
+
+import torch
+from transformers import AutoConfig
+
+from lmdeploy.tokenizer import Tokenizer
+from lmdeploy.utils import get_logger
+
+from ..config import CacheConfig, ModelConfig, SchedulerConfig
+from ..messages import (MessageStatus, SamplingParam, SchedulerSequence,
+                        SchedulerSession)
+from ..paging import Scheduler
+from .logits_process import FusedLogitsProcessor
+from .model_agent import BaseModelAgent, ModelInputs, TPModelAgent
+from .request import (Request, RequestManager, RequestType, Response,
+                      ResponseType)
+
+logger = get_logger('lmdeploy')
+
+
+@dataclass
+class InferOutput:
+    """The output of the model inference."""
+
+    session_id: int
+    token_ids: List[int]
+    sender_id: int
+    req_id: int
+    meta: Any = None
+    finish: bool = False
+    logits: torch.Tensor = None
+
+
+def _check_resp(resp: Response, state: ResponseType, warning_msg: str = None):
+    """check if response has state."""
+    ret = resp.type == state
+    if not ret and warning_msg is not None:
+        logger.warning(warning_msg)
+    return ret
+
+
+def _check_resp_success(resp: Response, warning_msg: str = None):
+    """check if response success."""
+    return _check_resp(resp, ResponseType.SUCCESS, warning_msg)
+
+
+def _get_torch_dtype(config: Any, default: str = 'float16'):
+    """Get the torch dtype from the model config.
+
+    Args:
+        config: Config of the hf model.
+        default (str): default device type.
+    """
+    torch_dtype = getattr(config, 'torch_dtype', default)
+    return eval(f'torch.{torch_dtype}')
+
+
+def _build_model_config(model_path: str, hf_config: Any):
```

Comment:
baichuan2 7b failed `python3 -m lmdeploy.pytorch.chat --model-name baichuan2-7b --model-path ./baichuan2-chat-7b`
```
Xformers is not installed correctly. If you want to use memory_efficient_attention to accelerate training use the following command to install Xformers
pip install xformers.
12/25 15:23:03 - lmdeploy - INFO - block num: 1430
WARNING: Can not find tokenizer.json. It may take long time to initialize the tokenizer.

double enter to end input >>> hello

<reserved_106>hello<reserved_107> <class 'list'>
Exception in thread Thread-1:

Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/tritonserver/lmdeploy/lmdeploy/pytorch/engine/engine.py", line 716, in loop
    step_tokens: Dict[int, InferOutput] = self.step(
  File "/opt/tritonserver/lmdeploy/lmdeploy/pytorch/engine/engine.py", line 519, in step
    output = self.model_agent.forward(inputs,
  File "/opt/tritonserver/lmdeploy/lmdeploy/pytorch/engine/model_agent.py", line 282, in forward
    output = model_forward(
  File "/opt/tritonserver/lmdeploy/lmdeploy/pytorch/engine/model_agent.py", line 208, in model_forward
    output = patched_model(
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "<string>", line 10, in wrap_forward
  File "/root/.cache/huggingface/modules/transformers_modules/baichuan2-chat-7b/modeling_baichuan.py", line 685, in forward
    outputs = self.model(
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
TypeError: forward() got an unexpected keyword argument 'position_ids'
```

Reply:
using the origin path is OK :`/workspace/share/baichuan2/Baichuan2-7B-Chat`. But the softlink path `./baichuan2-chat-7b` fails.